src/Steps/Release/MergeUpRelease.php

```diff
@@ -0,0 +1,181 @@
+<?php
+
+namespace SilverStripe\Cow\Steps\Release;
+
+use SilverStripe\Cow\Model\Release\LibraryRelease;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+
+/**
+ * Merge up released commits to the major branch and optionally remove x.y-release branch.
+ * E.g. if you just released 4.12.0, this will merge 4.12-release up to 4.12, and then 4.12
+ * up to 4 - and then optionally delete the 4.12-release branch.
+ */
+class MergeUpRelease extends ReleaseStep
+{
+    public function getStepName()
+    {
+        return 'mergeup';
+    }
+
+    public function run(InputInterface $input, OutputInterface $output)
+    {
+        $this->log($output, "Merging up all modules");
+
+        $this->mergeRecursive($output, $this->getReleasePlan(), $input->getOption('delete-release-branch') ?? false);
+
+        $this->log($output, "All modules merged up");
+    }
+
+    /**
+     * Merge a library and its children up to their major branches
+     *
+     * @param OutputInterface $output
+     * @param LibraryRelease $releasePlanNode Node in release plan being released
+     * @param bool $deleteBranch Whether to delete the release branch after merging
+     */
+    protected function mergeRecursive(OutputInterface $output, LibraryRelease $releasePlanNode, bool $deleteBranch)
+    {
+        $currentBranch = $releasePlanNode->getLibrary()->getBranch();
+
+        // Skip upgrade-only modules and anything not on an actual branch (e.g. using a pre-existing tag)
+        if ($releasePlanNode->getLibrary()->isUpgradeOnly() || !$currentBranch) {
+            return;
+        }
+        $currentLibrary = $releasePlanNode->getLibrary()->getName();
+
+        // Skip if it's not a proper release branch.
+        $currentBranchIsRelease = preg_match(
+            '/' . preg_quote(RewriteReleaseBranches::RELEASE_SUFFIX) . '$/',
+            $currentBranch
+        );
+        if (!$currentBranchIsRelease) {
+            $this->log(
+                $output,
+                "Warning: Current branch '$currentBranch' is not a release branch. "
```

Comment:
I'd be inclined to exit at this point. The user may have done something like manually switch a branch in their release dir and forgot to change it back.

If we simply skip here, then it's possible that we'll miss that an entire recipe wasn't merged up properly

I guess it does get complex if we've previously exited because there was an issue with any merge-ups where we'll exit, so, I do understand the motivation to skip here

I'm thinking perhaps we should move the -release branch deletion to its own step and do that at the very end.  That way we can reliably do detection at this point if there is an existing -release branch that we should possibly switch to?

Reply:
I'd be much happier exiting here than having deletion as its own separate command (I assume that's what you mean by "step" here?), so that's the change I'll make for now.
I can split off deletion as its own command if you really want, or if you did actually mean a separate step in the same command, then I'll need more information about what you meant and how you expect that flow to go.

> I guess it does get complex if we've previously exited because there was an issue with any merge-ups where we'll exit, so, I do understand the motivation to skip here

I don't think it does - after fixing issues with merge-ups, the release manager should checkout the release branch again before running the command again. So having this check is a good protection against forgetting to do that.

>  That way we can reliably do detection at this point if there is an existing -release branch that we should possibly switch to?

We could do that here anyway - but we will already be on the release branch, because that was the last thing to happen as part of the merging. I can put in another `if (!$releasePlanNode->isOnReleaseBranch())` check before deletion though if you like.

Reply:
> as its own separate command

yes that's what I meant.  To me it seems much safer to do -release branch deletion at the very end of the cow process, because it's technically an optional step and you could actually just leave those old branches lying around without much real world impact.  Retaining the -release branches during the release is completely safe, and they can be used within cow to detect if there was a release branch created, and if they differ from the dev branch.  This can be used as protection against things like "I manually changed to the dev branch from the release branch and forgot to change back"

Reply:
Done